napari/_qt/qt_debug_menu.py

```diff
@@ -0,0 +1,107 @@
+
+Current Items
+-------------
+Trace File -> Start Tracing...
+Trace File -> Stop Tracking
+"""
+from qtpy.QtCore import QTimer
+from qtpy.QtWidgets import QAction, QFileDialog
+
+from ..utils import perf
+
+
+def _ensure_extension(filename: str, extension: str):
+    """Add the extension if needed."""
+    if filename.endswith(extension):
+        return filename
+    return filename + extension
+
+
+class DebugMenu:
+    def __init__(self, main_window):
+        """Create the debug menu.
+
+        Parameters
+        ----------
+        main_menu : qtpy.QtWidgets.QMainWindow.menuBar
+            We add ourselves to this menu.
+        """
+        self.debug_menu = main_window.main_menu.addMenu('&Debug')
+
+        self.perf = PerformanceSubMenu(
+            main_window, self.debug_menu.addMenu("Performance Trace")
+        )
+
+
+class PerformanceSubMenu:
+    """The flyout menu to start/stop recording a trace file.
+    """
+
+    def __init__(self, main_window, sub_menu):
+        self.main_window = main_window
+        self.sub_menu = sub_menu
+        self.start = self._add_start()
+        self.stop = self._add_stop()
+        self._set_recording(False)
+
+    def _set_recording(self, recording: bool):
+        """Toggle which are enabled/disabled.
+
+        Parameters
+        ----------
+        record : bool
+            Are we currently recording a trace file.
+        """
+        self.start.setEnabled(not recording)
+        self.stop.setEnabled(recording)
+
+    def _add_start(self):
+        """Add Start Recording action.
+        """
+        start = QAction('Start Recording...', self.main_window._qt_window)
+        start.setShortcut('Alt+T')
+        start.setStatusTip('Start recording a trace file')
+        start.triggered.connect(self._start_trace)
+        self.sub_menu.addAction(start)
+        return start
+
+    def _add_stop(self):
+        """Add Stop Recording action.
+        """
+        stop = QAction('Stop Recording', self.main_window._qt_window)
+        stop.setShortcut('Shift+Alt+T')
+        stop.setStatusTip('Stop recording a trace file')
+        stop.triggered.connect(self._stop_trace)
+        self.sub_menu.addAction(stop)
+        return stop
+
+    def _start_trace(self):
+        """Start recording a trace file."""
+        viewer = self.main_window.qt_viewer
+
+        filename, _ = QFileDialog.getSaveFileName(
+            parent=viewer,
+            caption='Record performance trace file',
+            directory=viewer._last_visited_dir,
+            filter="Trace Files (*.json)",
+        )
+        if filename:
+            filename = _ensure_extension(filename, '.json')
+
+            def start_trace():
+                perf.timers.start_trace_file(filename)
+                self._set_recording(True)
+
+            # Schedule this to avoid bogus "MetaCall" event for the entire
+            # time the file dialog was up.
+            QTimer.singleShot(0, start_trace)
```

Comment:
this feels like a hack, how were you previously connecting the `start_trace` function and showing the dialog?   Could you use the `fileSelected` signal?

Reply:
Good idea but I just tried it, same problem you get a giant `MetaCall` event in the trace, covering the many seconds you had the dialog open. It must call that signal immediately in a nested way. This kind of ruins the trace file so we don't want it there.

This is a very specialized situation related to timing. When `_start_trace` is called `QApplicationWithTiming` has a timer running. If we turn on tracing immediately that super long timer goes into the trace file.

The very first thing I tried was build logic into PerfTimers to say "when you start a trace, don't really start tracing for a little bit" but that seemed messy. I think `singleShot` does exactly what we want, it's one line, it's a legal command. Plus this is in the debug menu.